Basic breadth-first pathfinding in cerebellum

JNotelddim opened the pull request:
Use breadth-first pathfinding to decide how to get places (reduces chances of snake getting stuck/closed-off by its own body.)

Write cerebellum test: very basic test that cerebellum finds paths appropriately.

JNotelddim committed:
Basic breadth-first pathfinding in cerebellum

Use breadth-first pathfinding to decide how to get places (reduces
chances of us getting stuck/closed-off by our own body.

Write cerebellum test: very basic test that cerebellum finds paths
appropriately

mattcasiro commented on src/brain.py:
```diff
@@ -15,42 +16,51 @@ def __init__(self, my_id: str, board: Board) -> None:
         self.my_id: str = my_id
         self.other_snakes: List[Snake] = self.board.get_other_snakes(my_id)
         self.me: Snake = next((snake for snake in self.board.snakes if snake.id == my_id))
+        self.cerebellum = Cerebellum(self, None)
```
You shouldn't need `None` here

- [x] Make second option in Cerebellum optional (default to `None`)

mattcasiro commented on src/brain.py:
```diff
@@ -15,42 +16,51 @@ def __init__(self, my_id: str, board: Board) -> None:
+                decision = next((move for move in moves_for_first_path_step if move in valid_moves), None)
         else:
-            loop_moves = self.follow_tail()#self.circle_perimeter()#self.follow_tail()
+            tail = self.me.coordinates[-1]
```
You use this on line 28 as well, if you're going to make it a variable you should do it up there and reference it both times.

- [x] move variable up

mattcasiro commented on src/brain.py:
```diff
@@ -95,105 +105,17 @@ def get_threatening_snakes_moves(self) -> List[Coordinate]:
-#        print("Reached unexpected condition in follow_wall")
-#        return "left"
-
     def get_moves_to(self, coord: Coordinate) -> List[str]:
```
Make todo to change the name to 'get_move directions'  maybe?
Actually, I think something like `get_string_for_move`

JNotelddim replied:
im not sure I agree that either of those are more representative of what it does than "get moves to _"
because like in battlesnake, a "move" is one of ["left", "right", "up", "down"], and it takes in a coordinate which it's giving move options to, and it may give more than one option if there are multiple ways to get there.
I think if it's an issue of it should only be returning one move, then there's more to it than just the title, yknow?

mattcasiro commented on src/cerebellum.py:
```diff
@@ -0,0 +1,114 @@
+
+class Cerebellum:
+    """Find paths to given coordinates."""
+    def __init__(self, brain, selected_algorithm: Optional[str]) -> None:
```
- [x] default second param to `None`

mattcasiro commented on src/cerebellum.py:
```diff
@@ -0,0 +1,114 @@
+    """Find paths to given coordinates."""
+    def __init__(self, brain, selected_algorithm: Optional[str]) -> None:
+        self.selected_algorithm: str = selected_algorithm if selected_algorithm is not None else 'breadth_first'
+        self.brain = brain
```
This creates a circular dependency. Pass the items you need in directly.

- [x] Don't pass in the brain to a submodule of brain

mattcasiro reviewed (changes requested):
Still need to look at the algorithm.

mattcasiro commented on src/cerebellum.py:
```diff
@@ -0,0 +1,114 @@
+        came_from: dict = {}
+        came_from[str(start)] = None
+
+        collision_coordinates = [coordinate for snake in self.board.snakes for coordinate in snake.coordinates]
```
I would create collision coordinates in `__init__` I think, and just reference them here. Or perhaps even better in their own function call that uses all the data passed in? We'll need it in a few places.

JNotelddim replied:
I can see why this could be valuable, but there are (or at least were) places in brain where collision_coordinates is modified to contain more/less coordinates (tails, head neighbours, etc), so I figured since collision_coordinates isn't necessarily the same everywhere it'd be simplest just to create it separately?

mattcasiro commented on src/cerebellum.py:
```diff
@@ -0,0 +1,114 @@
+                matched = True
+                break
+
+            valid_move_coordinatees = [coord for coord in current.get_neighbours() if self.board.is_coordinate_in_bounds(coord) and coord not in collision_coordinates]
```
This is the one thing we will have trouble doing if we don't pass the board class in. Thoughts? We could do the math ourselves internally here... or board could pass in a list of collision coordinates that include edges?

JNotelddim replied:
board doesn't have a reference to brain, so I'm just passing "me" and "board" into cerebellum instead of brain and as far as I can tell there's no circular dependency

mattcasiro commented on src/cerebellum.py:
```diff
@@ -0,0 +1,114 @@
+        while not unexplored.empty():
+            current = unexplored.get()
+
+            if any(coord.x == current.x and coord.y == current.y for coord in coordinates):
```
I think I'd prefer passing in a single coordinate, and then calling the search multiple times, but maybe that's not a great solution for breadth-first. If you prefer it this way, no problem.

JNotelddim replied:
I'm not sure I see the benefit -- but it could certainly be worth trying.
However, since this is working, and competition's tomorrow, maybe we compromise on a trello task to explore that option?

mattcasiro commented on src/coordinate.py:
```diff
@@ -76,3 +76,14 @@ def get_right(self) -> Coordinate:
         """Return the Coordinate to the left of this Coordinate."""
         return Coordinate((self.x - 1, self.y))
+
+    def get_neighbours(self) -> List[Coordinate]:
```
Nice, I'll need this too. Awesome :)

JNotelddim committed:
Fix circular dependency

brain had cerebellum which had brain which was bad.
instead, pass "me" (snake) and "board" to cerebellum so it can do its
stuff
